internal/prefixcollector/collector.go

```diff
@@ -0,0 +1,133 @@
+// limitations under the License.
+
+// Package prefixcollector contains excluded prefix collector, prefix sources and functions working with them
+package prefixcollector
+
+import (
+	"cmd-exclude-prefixes-k8s/internal/utils"
+	"context"
+	"io/ioutil"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/excludedprefixes"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/prefixpool"
+)
+
+// ExcludePrefixSource is source of excluded prefixes
+type ExcludePrefixSource interface {
+	Prefixes() []string
+}
+
+// ExcludePrefixCollector is service, collecting excluded prefixes from list of ExcludePrefixSource
+// and environment variable "EXCLUDED_PREFIXES"
+// and writing result to outputFilePath in yaml format
+type ExcludePrefixCollector struct {
+	notifyChan          chan struct{}
+	baseExcludePrefixes []string
+	outputFilePath      string
+	sources             []ExcludePrefixSource
+}
+
+// ExcludePrefixCollectorOption is functional option for ExcludePrefixCollector
+type ExcludePrefixCollectorOption func(*ExcludePrefixCollector)
+
+const (
+	// DefaultConfigMapNamespace is default namespace of kubernetes ConfigMap
+	DefaultConfigMapNamespace = "default"
+	defaultConfigMapName      = "nsm-config-volume"
+	outputFilePermissions     = 0600
+)
+
+// WithFilePath returns ExcludePrefixCollectorOption, that set filePath as collector's outputFilePath
+func WithFilePath(filePath string) ExcludePrefixCollectorOption {
+	return func(collector *ExcludePrefixCollector) {
+		collector.outputFilePath = filePath
+	}
+}
+
+// WithSources returns ExcludePrefixCollectorOption, that set sources as collector's prefix sources
+func WithSources(sources ...ExcludePrefixSource) ExcludePrefixCollectorOption {
+	return func(collector *ExcludePrefixCollector) {
+		collector.sources = sources
+	}
+}
+
+// NewExcludePrefixCollector creates ExcludePrefixCollector
+// and applies every ExcludePrefixCollectorOption to it
+func NewExcludePrefixCollector(ctx context.Context, prefixesFromEnv []string,
+	configMapNamespace string, notifyChan chan struct{},
+	options ...ExcludePrefixCollectorOption) *ExcludePrefixCollector {
+	collector := &ExcludePrefixCollector{
+		outputFilePath:      excludedprefixes.PrefixesFilePathDefault,
+		baseExcludePrefixes: utils.GetValidatedPrefixes(prefixesFromEnv),
+		notifyChan:          notifyChan,
+	}
+
+	for _, option := range options {
+		option(collector)
+	}
+
+	if collector.sources == nil {
+		collector.sources = []ExcludePrefixSource{
+			NewKubeAdmPrefixSource(ctx, notifyChan),
+			NewKubernetesPrefixSource(ctx, notifyChan),
+			NewConfigMapPrefixSource(ctx, notifyChan, defaultConfigMapName, configMapNamespace),
+		}
+	}
+
+	return collector
+}
+
+// Start - starts every source, then begin monitoring notifyChan.
+// Updates exclude prefix file after every notification.
+func (epc *ExcludePrefixCollector) Start() {
+	go func() {
```

Comment:
I think we can move goroutine spawning on top-level. In this case code readers will see where we are using async call.
```suggestion
```

Reply:
Moved.